### PR TITLE
BoxDecoration should clip its backgroundImage if shape is BoxShape.circle

### LINKED ISSUE #5306: Option to correctly clip BackgroundImage in CircleAvatar
It would be very useful to be able to clip the _BackgroundImage_ for a _CircleAvatar_.
This could be set as a parameter so that "pre-clipped" images don't need to be processes.

### LINKED ISSUE #7289: BoxDecoration with BoxShape.circle doesn't clip backgroundImage
A BoxDecoration widget with `shape: BoxShape.circle` doesn't clip its backgroundImage.

One consequence of this is that CircleAvatar doesn't clip its backgroundImage either. See https://github.com/flutter/flutter/issues/5306

Similarly, using CircleAvatar in UserAccountsDrawerHeader doesn't do what's expected. See https://github.com/flutter/flutter/issues/7191.

```
import 'package:flutter/material.dart';

class BoxDecorationBackgroundImageClipFail extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text('Image clip should be a circle'),
      ),
      body: new Center(
        child: new Container(
          width: 100.0,
          height: 100.0,
          decoration: new BoxDecoration(
            shape: BoxShape.circle,
            backgroundColor: Theme.of(context).primaryColor,
            backgroundImage: new BackgroundImage(
              image: new AssetImage('assets/cat.jpg'),
            ),
          ),
        ),
      ),

    );
  }
}

void main() {
  runApp(new MaterialApp(home: new BoxDecorationBackgroundImageClipFail()));
}
```

### PR DESCRIPTION
**Pull request by HansMuller**
Fixes #7289
Fixes #5306

### CODE REVIEW

File: packages/flutter/test/painting/decoration_test.dart [hunk 1]
```diff
@@ -99,4 +143,48 @@ void main() {
...
+      async.flushMicrotasks();
+      expect(onChangedCalled, isTrue);
+      boxPainter.paint(canvas, Offset.zero, imageConfiguration);
+
+      // We expect a clip - for the BoxShape.circle - to preceed
+      // the drawImageRect call.
+      List<Invocation> commands = canvas.invocations.where((Invocation invocation) {
+        return invocation.memberName == #clipPath || invocation.memberName == #drawImageRect;
+      }).toList();
+      expect(commands.length, equals(2));
```
Comment (Hixie) [inline packages/flutter/test/painting/decoration_test.dart]: fwiw, you don't need to say `equals(2)`, you can just say `2`.

File: packages/flutter/lib/src/painting/box_painter.dart [hunk 2]
```diff
@@ -1051,7 +1051,7 @@ class BoxDecoration extends Decoration {
   /// * If [backgroundColor] is null, this decoration does not paint a background color.
   /// * If [backgroundImage] is null, this decoration does not paint a background image.
   /// * If [border] is null, this decoration does not paint a border.
-  /// * If [borderRadius] is null, this decoration use more efficient background
+  /// * If [borderRadius] is null, this decoration uses more efficient background
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/box_painter.dart]: This seems to just be a bald-faced lie. We totally ignore the border radius when drawing the background image, as far as I can tell.

Maybe that's another bug? We don't clip the background image to the border radius?

**Comment by Hixie** [comment]
LGTM. Please file a bug about the border radius not clipping the background image.

File: packages/flutter/test/painting/decoration_test.dart [hunk 3]
```diff
@@ -146,45 +146,55 @@ void main() {
...
 
-      List<Invocation> invocations = <Invocation>[];
-      TestCanvas canvas = new TestCanvas(invocations);
-      ImageConfiguration imageConfiguration = const ImageConfiguration(
-          size: const Size(100.0, 100.0)
+      testClippedBoxDecoration(
+        new BoxDecoration(
+          borderRadius: new BorderRadius.all(const Radius.circular(16.0)),
+          backgroundImage: backgroundImage,
+        ),
```
Comment (Hixie) [inline packages/flutter/test/painting/decoration_test.dart]: i guess we should have a control test too that verifies there's no clip if it's not one of these...

**Comment by Hixie** [comment]
That's awesome. LGTM! :-)

File: packages/flutter/lib/src/painting/box_painter.dart (hunk 2 above)
Comment (HansMuller) [inline packages/flutter/lib/src/painting/box_painter.dart]: I added support for clipping the backgroundImage when borderRadius is specified.

screenshot

File: packages/flutter/test/painting/decoration_test.dart (hunk 1 above)
Comment (HansMuller) [inline packages/flutter/test/painting/decoration_test.dart]: Good point; done

File: packages/flutter/lib/src/painting/box_painter.dart [hunk 4]
```diff
@@ -1333,6 +1334,17 @@ class _BoxDecorationPainter extends BoxPainter {
...
+
+    Path clipPath;
+    if (_decoration.shape == BoxShape.circle)
+      clipPath = new Path()..addOval(rect);
+    else if (_decoration.borderRadius != null)
+      clipPath = new Path()..addRRect(_decoration.borderRadius.toRRect(rect));
+    if (clipPath != null) {
+      canvas.save();
+      canvas.clipPath(clipPath);
+    }
```
Comment (abarth) [inline packages/flutter/lib/src/painting/box_painter.dart]: Did we already hit test correct?

File: packages/flutter/test/painting/decoration_test.dart (hunk 3 above)
Comment (HansMuller) [inline packages/flutter/test/painting/decoration_test.dart]: Yes, good catch.  I've added one.

**Pull request merged**